### PR TITLE
[WebAssembly] Implement prototype f32.load_f16 instruction.

### PR DESCRIPTION
**Pull request by brendandahl**
Adds a builtin and intrinsic for the f32.load_f16 instruction.

The instruction loads an f16 value from memory and puts it in an f32. Specified at:
https://github.com/WebAssembly/half-precision/blob/29a9b9462c9285d4ccc1a5dc39214ddfd1892658/proposals/half-precision/Overview.md

Note: the current spec has f32.load_f16 as opcode 0xFD0120, but this is incorrect and will be changed to 0xFC30 soon.

### CODE REVIEW

**Comment by brendandahl** [comment]
/cc @tlively @dschuff
(I guess I can't assign reviewers since I don't have commit access.)

**Review by dschuff — APPROVED** [review]
Overall this looks good, and I think it makes sense to model this as short* for now. I think it will be interesting to see if that ends up causing issues. Out of curiosity does this work if you try `_fp16`?

File: llvm/include/llvm/IR/IntrinsicsWebAssembly.td [hunk 1]
```diff
@@ -321,6 +321,18 @@ def int_wasm_relaxed_dot_bf16x8_add_f32:
...
+// Half-precision intrinsics (experimental)
+//===----------------------------------------------------------------------===//
+
+// TODO: Replace these intrinsic with normal ISel patterns once the XXX
+// instructions are merged to the proposal.
+def int_wasm_loadf16_f32:
+  Intrinsic<[llvm_float_ty],
+            [llvm_ptr_ty],
+            [IntrReadMem, IntrArgMemOnly],
+             "", [SDNPMemOperand]>;
```
Comment (aheejin) [inline llvm/include/llvm/IR/IntrinsicsWebAssembly.td:334]: Why is the name empty?

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td [hunk 2]
```diff
@@ -38,6 +38,13 @@ multiclass RELAXED_I<dag oops_r, dag iops_r, dag oops_s, dag iops_s,
                             asmstr_s, simdop, HasRelaxedSIMD>;
 }
 
+multiclass HALF_PRECISION_I<dag oops_r, dag iops_r, dag oops_s, dag iops_s,
+                            list<dag> pattern_r, string asmstr_r = "",
+                            string asmstr_s = "", bits<32> simdop = -1> {
+  defm "" : ABSTRACT_SIMD_I<oops_r, iops_r, oops_s, iops_s, pattern_r, asmstr_r,
+                            asmstr_s, simdop, HasHalfPrecision>;
+}
+
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:47]: Is this used in this PR? If not, how about adding this when this is used?

**Review by aheejin — COMMENTED** [review]
Given that this uses `LOADPAT`, how about adding offset tests like other loads in https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/WebAssembly/offset.ll ?

File: llvm/include/llvm/IR/IntrinsicsWebAssembly.td (hunk 1 above)
Comment (brendandahl) [reply llvm/include/llvm/IR/IntrinsicsWebAssembly.td:334]: It looks like we have empty/missing names for nearly all of the intrinsics except for `int_wasm_ref_is_null_extern` and `int_wasm_ref_is_null_func`. Do you know what the name is used for? Maybe it's for when the name can't be automatically translated to an llvm name?

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td (hunk 2 above)
Comment (brendandahl) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:47]: This will be for my next PRs. I'll remove.

**Comment by brendandahl** [comment]
> Overall this looks good, and I think it makes sense to model this as short* for now. I think it will be interesting to see if that ends up causing issues. Out of curiosity does this work if you try `_fp16`?

I was trying _Float16 and that wasn't working since it requires the target to support it. `__fp16` does work though. I'll change it.

File: llvm/include/llvm/IR/IntrinsicsWebAssembly.td (hunk 1 above)
Comment (aheejin) [reply llvm/include/llvm/IR/IntrinsicsWebAssembly.td:334]: Hmm I'm not sure. If this works without the name I guess it's fine.

File: llvm/test/CodeGen/WebAssembly/offset.ll [hunk 3]
```diff
@@ -666,3 +666,29 @@ define {i32,i32,i32,i32} @aggregate_return() {
 define {i64,i32,i16,i8} @aggregate_return_without_merge() {
   ret {i64,i32,i16,i8} zeroinitializer
 }
+
+;===----------------------------------------------------------------------------
```
Comment (brendandahl) [inline llvm/test/CodeGen/WebAssembly/offset.ll:670]: I didn't add many tests here, since it seems pretty well covered for other load patters.

**Review by aheejin — APPROVED** [review]

**Pull request merged**